db/schema_tables.cc

```diff
@@ -1095,26 +1120,95 @@ future<> store_column_mapping(distributed<service::storage_proxy>& proxy, schema
+            return std::nullopt;
+        }
+        auto table_name = get_table_name(rt.start);
+        if (table_name != get_table_name(rt.end)) {
+            slogger.trace("Mutation of {}.{} for keyspace {} contains a multi-table range tombstone",
+                          m.schema()->ks_name(), m.schema()->cf_name(), keyspace_name);
+            return std::nullopt;
+        }
+        table_names.emplace(std::move(table_name));
+        break;
+    }
+    for (auto&& row : m.partition().clustered_rows()) {
+        table_names.emplace(get_table_name(row.key()));
+    }
+    slogger.trace("Mutation of {}.{} for keyspace {} affects tables: {}",
+                  m.schema()->ks_name(), m.schema()->cf_name(), keyspace_name, table_names);
+    return table_names;
+}
+
 static future<> do_merge_schema(distributed<service::storage_proxy>& proxy, std::vector<mutation> mutations, bool do_flush)
 {
     slogger.trace("do_merge_schema: {}", mutations);
     schema_ptr s = keyspaces();
     // compare before/after schemas of the affected keyspaces only
     std::set<sstring> keyspaces;
     std::set<table_id> column_families;
+    std::unordered_map<sstring, std::optional<std::set<sstring>>> affected_tables;
     for (auto&& mutation : mutations) {
-        keyspaces.emplace(value_cast<sstring>(utf8_type->deserialize(mutation.key().get_component(*s, 0))));
+        sstring keyspace_name = value_cast<sstring>(utf8_type->deserialize(mutation.key().get_component(*s, 0)));
+        keyspaces.emplace(keyspace_name);
+
+        if (schema_tables_holding_schema_mutations().contains(mutation.schema()->id())) {
+            auto i = affected_tables.find(keyspace_name);
+            if (i == affected_tables.end() || i->second) {
+                auto table_names = get_affected_tables(keyspace_name, mutation);
+                if (table_names && i != affected_tables.end()) {
+                    i->second->merge(*table_names);
+                } else {
+                    affected_tables[keyspace_name] = std::move(table_names);
+                }
+            }
+        }
+
         column_families.emplace(mutation.column_family_id());
         // We must force recalculation of schema version after the merge, since the resulting
         // schema may be a mix of the old and new schemas.
         delete_schema_version(mutation);
     }
 
+    for (auto&& [keyspace_name, maybe_tables] : affected_tables) {
+        if (!maybe_tables) {
+            slogger.trace("Reading table list for keyspace {}", keyspace_name);
+            std::set<sstring> table_names;
+            table_names = boost::copy_range<std::set<sstring>>(
+                    co_await read_table_names_of_keyspace(proxy, keyspace_name, tables()));
+            table_names.merge(boost::copy_range<std::set<sstring>>(
+                    co_await read_table_names_of_keyspace(proxy, keyspace_name, views())));
+            maybe_tables = std::move(table_names);
```

Comment:
There's a bug here. If maybe_tables is std::nullopt, we should still collect known affected tables from the mutations and add them on top of what is read here. Otherwise we may miss new tables in the post-apply phase.